consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ProposalValidator.java

```diff
@@ -0,0 +1,276 @@
+import org.hyperledger.besu.consensus.common.bft.BftBlockInterface;
+import org.hyperledger.besu.consensus.common.bft.ConsensusRoundIdentifier;
+import org.hyperledger.besu.consensus.common.bft.payload.Payload;
+import org.hyperledger.besu.consensus.common.bft.payload.SignedData;
+import org.hyperledger.besu.consensus.qbft.messagewrappers.Proposal;
+import org.hyperledger.besu.consensus.qbft.payload.PreparePayload;
+import org.hyperledger.besu.consensus.qbft.payload.PreparedRoundMetadata;
+import org.hyperledger.besu.consensus.qbft.payload.RoundChangePayload;
+import org.hyperledger.besu.ethereum.BlockValidator;
+import org.hyperledger.besu.ethereum.ProtocolContext;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.Hash;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ProposalValidator {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private static final String ERROR_PREFIX = "Invalid Proposal Payload";
+
+  private final BlockValidator blockValidator;
+  private final ProtocolContext protocolContext;
+  private final int quorumMessageCount;
+  private final Collection<Address> validators;
+  private final ConsensusRoundIdentifier roundIdentifier;
+  private final Address expectedProposer;
+
+  public ProposalValidator(
+      final BlockValidator blockValidator,
+      final ProtocolContext protocolContext,
+      final int quorumMessageCount,
+      final Collection<Address> validators,
+      final ConsensusRoundIdentifier roundIdentifier,
+      final Address expectedProposer) {
+    this.blockValidator = blockValidator;
+    this.protocolContext = protocolContext;
+    this.quorumMessageCount = quorumMessageCount;
+    this.validators = validators;
+    this.roundIdentifier = roundIdentifier;
+    this.expectedProposer = expectedProposer;
+  }
+
+  public boolean validate(final Proposal msg) {
+
+    final ProposalPayloadValidator payloadValidator =
+        new ProposalPayloadValidator(
+            expectedProposer, roundIdentifier, blockValidator, protocolContext);
+
+    if (!payloadValidator.validate(msg.getSignedPayload())) {
+      LOG.info("{}: invalid proposal payload in proposal message", ERROR_PREFIX);
+      return false;
+    }
+
+    if (!validateProposalAndRoundChangeAreConsistent(msg)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean validateProposalAndRoundChangeAreConsistent(final Proposal proposal) {
+    final ConsensusRoundIdentifier proposalRoundIdentifier = proposal.getRoundIdentifier();
+
+    if (proposalRoundIdentifier.getRoundNumber() == 0) {
+      return validateRoundZeroProposalHasNoRoundChangesOrPrepares(proposal);
+    } else {
+
```

Comment:
If all the roundChanges have empty prepareRound and prepareValue then check is the same as the round zero 0. Don't think we are doing this i.e.
```
proposedBlock.header.proposer == expectedProposer
&& validateEthereumBlock(blockchain, proposedBlock)
&& |validators(blockchain + [proposedBlock])| > 0
&& proposedBlock.header.timestamp >= blockchain[|blockchain|-1].header.timestamp + blockTime
```

Reply:
added checks to round-0 and also if no-prepared-metadata exists, to ensure the proposal author matches the coinbase .... but its a bit of a naff check, given this doesn't hold true on a prepared-roundchange.